### PR TITLE
Update database handling when switching between different databases

### PR DESCRIPTION
**Pull request by frzyc**
- Allow databases to switch their storage, while maintaining their cached data.
- Database swapping no long have expensive new database initialization steps, since they are swapped "hot".
- Use the data entry classes in database to set/unset data in storage, thus extra data in the localstorage are no longer being bundled in the data of other databases.

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -58,8 +58,21 @@ function ScrollTop({ children }: { children: React.ReactElement }) {
...
-  const dbContextObj = useMemo(() => ({ database, setDatabase }), [database, setDatabase])
+  const [databases, setDatabases] = useState(() => [1, 2, 3, 4].map(index => {
+    const curDBIndex = parseInt(localStorage.getItem("dbIndex") || "1")
+    if (index === curDBIndex) {
+      return new ArtCharDatabase(new DBLocalStorage(localStorage))
+    } else {
+      const dbName = `extraDatabase_${index}`
+      const eDB = localStorage.getItem(dbName)
+      const dbObj = eDB ? JSON.parse(eDB) : { dbIndex: `${index}` }
+      const db = new ArtCharDatabase(new SandboxStorage(dbObj))
```
Comment (nguyentvan7) [inline src/App.tsx:69]: Missing import for SandboxStorage

File: src/PageSettings/DatabaseCard.tsx [hunk 2]
```diff
@@ -115,34 +77,19 @@ function DataCard({ index, databaseContextObj }: { index: number, databaseContex
...
-    databaseContextObj.database.storage.copyFrom(database.storage)
-    databaseContextObj.database.storage.set("dbIndex", index)
-    databaseContextObj.setDatabase(new ArtCharDatabase(databaseContextObj.database.storage))
-
-  }, [databaseContextObj, database, index])
+    if (current) return
+    mainDB.toExtraLocalDB()
+    database.swapStorage(mainDB)
+    setDatabase(database)
+  }, [setDatabase, mainDB, database])
```
Comment (nguyentvan7) [inline src/PageSettings/DatabaseCard.tsx:84]: Missing dependency on `current`

File: src/PageSettings/UploadCard.tsx [hunk 3]
```diff
@@ -155,16 +156,19 @@ function MergeResult({ result, dbTotal, type }: { result: ImportResultCounter<an
...
-    database.storage.copyFrom(importedDatabase.storage)
-    setDatabase(new ArtCharDatabase(database.storage))
+    importedDatabase.swapStorage(database)
+    const dbs = [...databases]
+    dbs[index] = importedDatabase
+    setDatabases(dbs)
+    importedDatabase.toExtraLocalDB()
     reset()
-  }, [database, importedDatabase, reset, setDatabase])
+  }, [database, importedDatabase, reset, setDatabases])
```
Comment (nguyentvan7) [inline src/PageSettings/UploadCard.tsx:171]: Missing dependency on `databases` and `index`

**Review by nguyentvan7 — CHANGES_REQUESTED** [review]

File: src/Database/migrate.ts [hunk 4]
```diff
@@ -94,17 +94,16 @@ export function migrate(storage: DBStorage) {
...
-  })
+  migrateVersion(8, () => { })
 
   // 6.1.9 - 6.2.3
   migrateVersion(11, () => {
     for (const key of storage.keys) {
       if (key.startsWith("char_")) {
         const character = storage.get(key)
-        const { baseStatOverrides = {} } = character
+        if (!character) continue
```
Comment (lantua) [inline src/Database/migrate.ts:104]: Should already exist from `key` being in `storage.keys`, use `storage.get(key)!`.

**Comment by nguyentvan7** [comment]
1. Upload an existing database
2. Swap to it
3. Delete it
4. Upload another database
5. You now have no database selected visually on the screen

**Comment by nguyentvan7** [comment]
swap button is completely broken on my end now. doesnt do anything

**Comment by frzyc** [comment]
@nguyentvan7 database swapping is working fine on my end. Can you double check?

**Comment by nguyentvan7** [comment]
Database numbers appear to be 0-indexed now. They are 1-indexed on prod

**Review by nguyentvan7 — APPROVED** [review]

**Pull request merged**